rdr_service/genomic/genomic_job_components.py

```diff
@@ -424,6 +428,126 @@ def _ingest_gc_manifest(self, data, _site):
         except RuntimeError:
             return GenomicSubProcessResult.ERROR
 
+    def ingest_single_aw1_row_for_member(self, member):
+        # Open file and pull row based on member.biobankId
+        with self.controller.storage_provider.open(self.target_file, 'r') as aw1_file:
+            reader = csv.DictReader(aw1_file, delimiter=',')
+            row = [r for r in reader if r['Biobank Id'] == str(member.biobankId)][0]
+
+            # Alter field names to remove spaces and change to lower case
+            row = dict(zip([key.lower().replace(' ', '').replace('_', '')
+                       for key in row], row.values()))
+
+        ingested_before = member.reconcileGCManifestJobRunId is not None
+
+        # Write AW1 data to genomic_set_member table
+        gc_manifest_column_mappings = self.get_aw1_manifest_column_mappings()
+
+        # Set attributes from file
+        for key in gc_manifest_column_mappings.keys():
+            try:
+                member.__setattr__(key, row[gc_manifest_column_mappings[key]])
+            except KeyError:
+                member.__setattr__(key, None)
+
+        # Set other fields not in AW1 file
+        member.reconcileGCManifestJobRunId = self.job_run_id
+        member.aw1FileProcessedId = self.file_obj.id
+        member.gcSite = self._get_site_from_aw1()
+
+        # Only update the member's state if it was AW0
+        if member.genomicWorkflowState == GenomicWorkflowState.AW0:
+            member.genomicWorkflowState = GenomicWorkflowState.AW1
+            member.genomicWorkflowStateModifiedTime = clock.CLOCK.now()
+
+        # Update member in DB
+        self.member_dao.update(member)
+
+        # Update AW1 manifest record count
+        if not ingested_before and not self.controller.bypass_record_count:
+            self.increment_manifest_file_record_count_from_id()
+
+    def ingest_single_aw2_row_for_member(self, member):
+        # Open file and pull row based on member.biobankId
+        with self.controller.storage_provider.open(self.target_file, 'r') as aw1_file:
+            reader = csv.DictReader(aw1_file, delimiter=',')
+            row = [r for r in reader if r['Biobank ID'] == str(member.biobankId)][0]
+
+            # Alter field names to remove spaces and change to lower case
+            row = dict(zip([key.lower().replace(' ', '').replace('_', '')
+                            for key in row], row.values()))
+
+        row['member_id'] = member.id
+        row['file_id'] = self.file_obj.id
+
+        # Truncate call rate
+        try:
+            row['callrate'] = row['callrate'][:10]
+        except KeyError:
+            pass
+
+        # Set contamination field to fload
+        row['contamination'] = float(row['contamination'])
+
+        # Percentages shouldn't be less than 0
+        if row['contamination'] < 0:
+            row['contamination'] = 0
```

Comment:
From a maintenance/reusability perspective: there's code elsewhere that runs through a file and does this kind of processing for each row in the file, right? Is the only difference is that it is coupled to code that uses it for multiple rows in a file? What do you think of reusing that code (and refactoring it a bit if needed)? (that way, if anything changes about how to process the file, there's only one place to make updates)

Reply:
Good point, I meant to address this once I got the code working. The major differences between the new and existing code are that the existing code checks for control samples, processes multiple records, and needs to do a lookup on the `genomic_set_member object`. 
I can cut down on the duplicated code by refactoring the repeated code sections into their own functions.